File: ShowControl/FundingCAPTCHA/touch-input.js
/**
 * TouchInput — maps world-space blob positions to grid tap events.
 *
 * Pure processor: feed it blob arrays via update(), get onTap() callbacks.
 * No WebSocket, no DOM.
 *
 * Coordinate convention: X=right, Z=up (height on vertical screen).
 * Row 0 is the top of the screen (high Z); row (rows-1) is the bottom (low Z).
 */
class TouchInput {
  /**
   * @param {object} opts
   * @param {number} opts.cols
   * @param {number} opts.rows
   * @param {{ x0: number, x1: number, z0: number, z1: number }} opts.screenRect
   *   World-space bounding box (cm) of the screen surface.
   *   x0/x1 = left/right edges; z0/z1 = bottom/top edges.
   * @param {number} [opts.dwellFrames=3]  Frames a blob must stay in a cell before tap fires.
   * @param {(col: number, row: number) => void} opts.onTap
   */
  constructor({ cols, rows, screenRect, dwellFrames = 3, onTap }) {
    this._cols        = cols;
    this._rows        = rows;
    this._screenRect  = screenRect;
    this._dwellFrames = dwellFrames;
    this._onTap       = onTap;
    this._dwell       = new Map(); // blobId → { col, row, frames }
  }

  /** Update screenRect and/or dwellFrames without recreating the instance. */
  configure({ screenRect, dwellFrames } = {}) {
    if (screenRect  !== undefined) this._screenRect  = screenRect;
    if (dwellFrames !== undefined) this._dwellFrames = dwellFrames;
  }

  /** Change grid dimensions and clear all in-progress dwell state. */
  resize(cols, rows) {
    this._cols = cols;
    this._rows = rows;
    this._dwell.clear();
  }

  /**
   * Process one frame of blobs.
   * @param {{ id: number, x: number, y: number, z: number }[]} blobs
   *   World-space positions in cm. y (depth) is ignored for grid mapping.
   */
  update(blobs) {
    const liveIds = new Set(blobs.map(b => b.id));

    for (const id of this._dwell.keys()) {
      if (!liveIds.has(id)) this._dwell.delete(id);
    }

    for (const b of blobs) {
      const { col, row } = this._toCell(b.x, b.z);
      const prev = this._dwell.get(b.id);

      if (!prev || prev.col !== col || prev.row !== row) {
        this._dwell.set(b.id, { col, row, frames: 1 });
      } else {
        prev.frames++;
      }
      if (this._dwell.get(b.id).frames === this._dwellFrames) {
        this._onTap(col, row);
      }
    }
  }

  /**
   * Convert world-space (x, z) to (col, row).
   * X increases rightward → col increases.
   * Z increases upward → row *decreases* (row 0 = top of screen).
   */
  _toCell(x, z) {
    const { x0, x1, z0, z1 } = this._screenRect;
    const col = Math.floor((x  - x0) / (x1 - x0) * this._cols);
    const row = Math.floor((z1 - z)  / (z1 - z0) * this._rows);
    return {
      col: Math.max(0, Math.min(this._cols - 1, col)),
      row: Math.max(0, Math.min(this._rows - 1, row)),
    };
  }
}

if (typeof module !== 'undefined') module.exports = { TouchInput };

File: ShowControl/FundingCAPTCHA/touch_input.py
"""
Blob-to-grid touch input for pygame games.

Wraps ScreenProjector + DwellTracker from touch_calibration.py.
Call update(blobs) each frame; on_tap(col, row) fires once per dwell.
"""
from __future__ import annotations

from touch_calibration import DwellTracker, ScreenProjector


class TouchInput:
    def __init__(
        self,
        cols: int,
        rows: int,
        screen_corners: dict,
        dwell_frames: int = 3,
        max_plane_dist_cm: float = 30.0,
        on_tap=None,
    ) -> None:
        self._cols           = cols
        self._rows           = rows
        self._dwell_frames   = dwell_frames
        self._max_dist       = max_plane_dist_cm
        self.on_tap          = on_tap  # callable(col, row) — set by orchestrator
        self._projector      = ScreenProjector(
            screen_corners["bottom_left"],
            screen_corners["bottom_right"],
            screen_corners["top_left"],
        )
        self._dwell          = DwellTracker(dwell_frames)
        self._prev_flashing: set[tuple[int, int]] = set()

    def resize(self, cols: int, rows: int) -> None:
        self._cols  = cols
        self._rows  = rows
        self._dwell = DwellTracker(self._dwell_frames)
        self._prev_flashing = set()

    def update(self, blobs: list[dict]) -> None:
        live: list[tuple[int, int, int]] = []
        for b in blobs:
            xyz = [b["x"], b["y"], b["z"]]
            if not self._projector.in_bounds_3d(xyz, self._max_dist):
                continue
            u, v = self._projector.project(xyz)
            if not ScreenProjector.in_bounds(u, v):
                continue
            col, row = ScreenProjector.uv_to_cell(u, v, self._cols, self._rows)
            live.append((int(b["id"]), col, row))

        prev = self._prev_flashing
        self._dwell.update(live)
        self._prev_flashing = self._dwell.flashing.copy()

        if self.on_tap:
            for cell in self._prev_flashing - prev:
                self.on_tap(cell[0], cell[1])
